broadcast/internal/composite/broadcaster.go

```diff
@@ -31,6 +34,120 @@ func NewBroadcaster(strategy Strategy, factories ...BroadcastFactory) broadcast.
+        }
+    }
+
+    result, err := c.strategy.Execute(ctx, executionFuncs)
+    if err != nil {
+        return nil, err
+    }
+
+    policyQuoteResponse, ok := result.(*broadcast.PolicyQuoteResponse)
+    if !ok {
+        return nil, fmt.Errorf("unexpected result type: %T", result)
+    }
+
+    return policyQuoteResponse, nil
+}
+
+func (c *compositeBroadcaster) GetFeeQuote(ctx context.Context) (*broadcast.FeeQuote, error){
+    executionFuncs := make([]executionFunc, len(c.broadcasters))
+    for i, broadcaster := range c.broadcasters {
+        executionFuncs[i] = func(ctx context.Context) (Result, error) {
+            return broadcaster.GetFeeQuote(ctx)
+        }
+    }
+
+    result, err := c.strategy.Execute(ctx, executionFuncs)
+    if err != nil {
+        return nil, err
+    }
+
+    feeQuote, ok := result.(*broadcast.FeeQuote)
+    if !ok {
+        return nil, fmt.Errorf("unexpected result type: %T", result)
+    }
+
+    return feeQuote, nil
+}
+
+func (c *compositeBroadcaster) GetBestQuote(ctx context.Context) (*broadcast.FeeQuote, error) {
+    fees := make([]*broadcast.FeeQuote, len(c.broadcasters))
+    for i, broadcaster := range c.broadcasters {
+        feeQuote, err := broadcaster.GetFeeQuote(ctx)
+        if err != nil {
+            fees[i] = nil
+        } else {
+            fees[i] = feeQuote
+        }
+    }
+
+    var bestQuote *broadcast.FeeQuote = nil
+    for _, feeQuote := range fees {
+        if feeQuote != nil {
+            bestQuote = feeQuote
+            break
+        }
+    }
+    if bestQuote == nil {
+        return nil, errors.New("failed to get reponse from any miner")
+    }
+
+    for _, feeQuote := range fees {
+        if feeQuote == nil {
+            continue
+        }
+        feePer1000Bytes := (1000 / feeQuote.MiningFee.Bytes) * feeQuote.MiningFee.Satoshis
+        bestFeePer1000Bytes := (1000 / bestQuote.MiningFee.Bytes) * bestQuote.MiningFee.Satoshis
+        if feePer1000Bytes < bestFeePer1000Bytes {
+            bestQuote = feeQuote
+        }
+    }
+
+    return bestQuote, nil
+}
+
+func (c *compositeBroadcaster) GetFastestQuote(ctx context.Context, timeout time.Duration) (*broadcast.FeeQuote, error) {
+	if timeout.Seconds() == 0 {
+		timeout = broadcast.DefaultFastestQuoteTimeout
+	}
+
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+    feeChannel := make(chan *broadcast.FeeQuote, len(c.broadcasters))
+
+	var wg sync.WaitGroup
+    for _, broadcaster := range c.broadcasters {
+        wg.Add(1)
+        go func (ctxWithTimeout context.Context, broadcaster broadcast.Client)  {
+            defer wg.Done()
+            feeQuote, err := broadcaster.GetFeeQuote(ctxWithTimeout)
+            if err == nil {
+                feeChannel <- feeQuote
+            }
+        } (ctxWithTimeout, broadcaster)
+    }
+
+    go func() {
+        wg.Wait()
+        close(feeChannel)
+    }()
+
+    fastQuote := <-feeChannel
+
+    if fastQuote == nil {
```

Comment:
As I understand it, we should wait for the fastest correct quote, so we should make sure `nil` never enters the channel

Reply:
Nil will enter the channel when all miners will timeout or throw an error (or some miners will timeout and the rest will throw an error), so basically when there is zero responses from all the queries.

Reply:
Yes, you're right. My bad ;-)